### PR TITLE
ScalarPairs are offset==0 field + other non-zst field

### LINKED ISSUE #51300: Constant evaluation error: attempted to read undefined bytes
I saw that #51086 was closed when #51225 was merged so I thought it would fix the same error on my own code as well however as of nightly: `rustc 1.28.0-nightly (aa094a43c 2018-06-01)` it still happens.

Code:
```rust
#[derive(PartialEq, Eq, Clone, Copy)]
pub struct Stat {
    pub id: u8,
    pub index: usize,
}

impl Stat {
    pub const STUDENT_HAPPINESS: Stat = Stat{
        id: 0,
        index: 0,
    };
    pub const STUDENT_HUNGER: Stat = Stat{
        id: 0,
        index: Self::STUDENT_HAPPINESS.index + 1,
    };

}

pub fn from_index(id: u8, index: usize) -> Option<Stat> {
    let stat = Stat{id, index};
    match stat {
        Stat::STUDENT_HAPPINESS => Some(Stat::STUDENT_HAPPINESS),
        Stat::STUDENT_HUNGER => Some(Stat::STUDENT_HUNGER),
        _ => None,
    }
}

fn main() { }

```
Playground: http://play.rust-lang.org/?gist=cf51c77c20f06ca15f1dc2cc4fa779b0&version=nightly&mode=debug

From what I can tell its an alignment issue because changing id from a `u8` to a `u64` fixes the issue.

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb 

fixes #51300

### CODE REVIEW

File: src/librustc_mir/interpret/place.rs [hunk 1]
```diff
@@ -137,7 +137,7 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
             Value::Scalar(_) if offset.bytes() == 0 && field.size == base_layout.size => Ok(Some((base, field.ty))),
             // split fat pointers, 2 element tuples, ...
             Value::ScalarPair(a, b) if base_layout.fields.count() == 2 => {
-                let val = [a, b][field_index];
+                let val = [a, b][base_layout.fields.memory_index(field_index)];
```
Comment (eddyb) [inline src/librustc_mir/interpret/place.rs]: This is still not a good idea (if you want to see what happens, try adding a bunch of ZSTs) - what you want is to check whether `offset.bytes() == 0` or not. I could've sworn I already wrote the code in this module to do it correctly, are there several similar pieces of code in here?

File: src/librustc_mir/interpret/terminator/mod.rs [hunk 2]
```diff
@@ -388,11 +388,12 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
...
                                             match other {
-                                                Value::Scalar(value) | Value::ScalarPair(value, _) => write_next(value)?,
+                                                Value::Scalar(value) => write_next(value)?,
+                                                Value::ScalarPair(a, b) => write_next([a, b][layout.fields.memory_index(0)])?,
                                                 _ => unreachable!(),
                                             }
-                                            if let Value::ScalarPair(_, value) = other {
-                                                write_next(value)?;
+                                            if let Value::ScalarPair(a, b) = other {
+                                                write_next([a, b][layout.fields.memory_index(1)])?;
```
Comment (eddyb) [inline src/librustc_mir/interpret/terminator/mod.rs]: Same here.

File: src/librustc_mir/transform/const_prop.rs [hunk 3]
```diff
@@ -221,7 +221,7 @@ impl<'b, 'a, 'tcx:'b> ConstPropagator<'b, 'a, 'tcx> {
                             trace!("layout computed");
                             use rustc_data_structures::indexed_vec::Idx;
                             let field_index = field.index();
-                            let val = [a, b][field_index];
+                            let val = [a, b][base_layout.fields.memory_index(field_index)];
```
Comment (eddyb) [inline src/librustc_mir/transform/const_prop.rs]: Same here.

File: src/librustc_mir/interpret/terminator/mod.rs [hunk 4]
```diff
@@ -338,65 +338,39 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
...
+                                            // this is the case where the field covers the entire type
+                                            assert_eq!(offset.bytes(), 0);
+                                            other
+                                        },
+                                        Value::ScalarPair(a, _) if offset.bytes() == 0 => Value::Scalar(a),
+                                        Value::ScalarPair(_, b) => Value::Scalar(b),
+                                        Value::Scalar(_) => bug!("Scalar does not cover entire type"),
+                                    };
+                                    let dest =
+                                        self.eval_place(&mir::Place::Local(arg_local))?;
```
Comment (eddyb) [inline src/librustc_mir/interpret/terminator/mod.rs]: Why not do something like this for the source, too?

File: src/librustc_mir/transform/const_prop.rs [hunk 5]
```diff
@@ -214,24 +214,10 @@ impl<'b, 'a, 'tcx:'b> ConstPropagator<'b, 'a, 'tcx> {
...
-                                param_env: self.param_env,
-                            };
-                            let field = base_layout.field(cx, field_index).ok()?;
-                            trace!("projection resulted in: {:?}", val);
-                            Some((Value::Scalar(val), field.ty, span))
-                        },
-                        _ => None,
-                    }
+                    let (value, field_ty) = self.use_ecx(span, |this| {
+                        this.ecx.read_field(base, None, field, ty)
```
Comment (eddyb) [inline src/librustc_mir/transform/const_prop.rs]: Like this, maybe?

File: src/librustc_mir/interpret/place.rs [hunk 6]
```diff
@@ -136,8 +136,8 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
             Value::ScalarPair(..) |
             Value::Scalar(_) if offset.bytes() == 0 && field.size == base_layout.size => Ok(Some((base, field.ty))),
             // split fat pointers, 2 element tuples, ...
```
Comment (eddyb) [inline src/librustc_mir/interpret/place.rs]: This comment is too specific.

**Comment by oli-obk** [comment]
all builders took less than 2h except for the dist builder, which timed out after 3h :confused:

**Pull request merged**